lib/knapsack_pro/runners/test_unit_runner.rb

```diff
@@ -11,21 +11,32 @@ def self.run(args)
         if runner.test_files_to_execute_exist?
           adapter_class.verify_bind_method_called
 
-          require 'test/unit'
-
           cli_args =
             (args || '').split +
             runner.test_file_paths.map do |f|
               File.expand_path(f)
             end
 
-          exit ::Test::Unit::AutoRunner.run(
+          exit test_unit_autorunner_run(
             true,
             runner.test_dir,
             cli_args
           )
         end
       end
+
+      private
+
+      # https://www.rubydoc.info/github/test-unit/test-unit/Test/Unit/AutoRunner#run-class_method
+      def self.test_unit_autorunner_run(force_standalone, default_dir, argv)
+        require 'test/unit'
```

Comment:
Method `test_unit_autorunner_run` has been extracted to not load `require 'test/unit'` because this breaks the test environment for RSpec and causes random errors.

Reply:
Not a fan of the `require` inside a method tbh. :/ Also a bit worried that the test environment dictates how we organize the code containing the actual business logic.

What types of random errors have you encountered? Any idea as to why it worked previously? Was there a reason for the `require` to previously lie inside of the `if` statement instead of at the top of the file (where you would typically expect it)? Was this done to avoid other errors maybe?

Have you considered any other solutions to solve the problems you encountered?

Reply:
knapsack_pro gem supports multiple test runners in a single gem. But we do not list them as [dependencies](https://github.com/KnapsackPro/knapsack_pro-ruby/blob/f11fd4a411ff1ef36b37d66f0c70fdc3bd2cea69/knapsack_pro.gemspec#L28) in `knapsack_pro.gemspec`. They are listed there only as a [dev dependency](https://github.com/KnapsackPro/knapsack_pro-ruby/blob/f11fd4a411ff1ef36b37d66f0c70fdc3bd2cea69/knapsack_pro.gemspec#L31) for a testing purpose of knapsack_pro gem.

For instance, if someone is using only RSpec in his project then when he installs knapsack_pro it would be pointless to download all test runners like cucumber, test-unit, etc.

That is why we only require a particular test runner during the runtime of the gem when a method like  `KnapsackPro::Runners::TestUnitRunner.run` is called. Basically, when someone on purpose wants to run knapsack_pro for a given test runner like test-unit then we require `require 'test/unit'` and we assume the user already has test-unit gem installed in his Gemfile.


In the test environment for knapsack_pro gem we should not load multiple test runners at once because they impact each other. For instance `require 'test/unit'` breaks RSpec and a lot of weird errors were happening (state of tests was not clean up between test examples) https://app.circleci.com/pipelines/github/KnapsackPro/knapsack_pro-ruby/411/workflows/927f821d-df4f-4213-90d4-cc38c189cfa8/jobs/1398

That is why I extracted `test_unit_autorunner_run` method to hide `require 'test/unit'` and not call it in test runtime.

Why it was working previously? I guess I did not refresh my local copy of Gemfile.lock for a few years and when I updated it and new test-unit gem and new rspec  gem versions were installed then problem was revealed.

Reply:
I see, thanks for the extended explanation!